File: autosegmenter/src/diamond/diamond.cpp
/*
 * DermShare Autosegmenter
 *
 * Copyright (c) 2012-2015 Carnegie Mellon University
 * All rights reserved.
 *
 * This software is distributed under the terms of the Eclipse Public
 * License, Version 1.0 which can be found in the file named LICENSE.
 * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS SOFTWARE CONSTITUTES
 * RECIPIENT'S ACCEPTANCE OF THIS AGREEMENT
 */

#include <stdio.h>
#include <string.h>
#include <sstream>
#include <lib_filter.h>
#include "diamond.h"

using namespace std;

void CDiamondContext::write_attr(const char *name, const char *value)
{
	_attrs[name] = value;
}

void CDiamondContext::write_attr(const char *name, int value)
{
	stringstream s;
	s << value;
	write_attr(name, s.str().data());
}

void CDiamondContext::write_attr(const char *name, double value)
{
	stringstream s;
	s << value;
	write_attr(name, s.str().data());
}

void CDiamondContext::save(const char *filename)
{
	FILE *fp = fopen(filename, "w");
	if (!fp) {
		return;
	}
	for (map<string, string>::iterator it = _attrs.begin();
			it != _attrs.end(); it++) {
		fprintf(fp, "\"%s\"=\"%s\"\n", it->first.c_str(),
				it->second.c_str());
	}
	fclose(fp);
}

void CDiamondContext::save(lf_obj_handle_t ohandle, bool omit)
{
	for (map<string, string>::iterator it = _attrs.begin();
			it != _attrs.end(); it++) {
		string name = it->first;
		lf_write_attr(ohandle, name.c_str(),
				it->second.size() + 1,
				it->second.c_str());
		if (omit) {
			lf_omit_attr(ohandle, name.c_str());
		}
	}
}

File: autosegmenter/autosegment/diamond.py
#
# DermShare Autosegmenter
#
# Copyright (c) 2015,2016 Carnegie Mellon University
# All rights reserved.
#
# This software is distributed under the terms of the Eclipse Public
# License, Version 1.0 which can be found in the file named LICENSE.
# ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS SOFTWARE CONSTITUTES
# RECIPIENT'S ACCEPTANCE OF THIS AGREEMENT

import numpy as np

from opendiamond.filter import Filter
from PIL import Image
from skimage.transform import rescale

from .util import watchdog, make_boring
from .image import resize
from .hairremover import HairRemover
from .segmenter import Segmenter

TIMEOUT = 300
DEFAULT_IMAGE_SIZE = (640, 640)
HEATMAP_NAME = '_filter.%s.heatmap.png'

def processOneImage(image, debug=None):
    # set a 5 minute execution timeout
    with watchdog(TIMEOUT):
        make_boring()

        image, scale = resize(image, DEFAULT_IMAGE_SIZE)
        if debug:
            debug['scaled_original'] = image

        print "=================\nRemoving hair ..."
        hair_removed = HairRemover(image, debug)

        print "=================\nSegmenting ..."
        segmentation = Segmenter(hair_removed, debug)
        return rescale(segmentation, 1./scale, order=3)

    return None

class AutoSegmentationFilter(Filter):
    def __call__(self, obj):
        results = processOneImage(obj.image)

        segmentation = results['segmentation']
        segmentation = Image.fromarray(np.uint8(segmentation * 255.))
        obj.set_heatmap(HEATMAP_NAME % self.session.name, segmentation)
        return True

